### PR TITLE
Refactor tokenizer types file

### PR DESCRIPTION
**Pull request by xtuc**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | no
| Deprecations?     | no
| Spec compliancy?  | no
| Tests added/pass? | no
| Fixed tickets     | no
| License           | MIT

This is not particularly wanted but it seems clearer to me. It's also to be consistent in usage / definition.

### CODE REVIEW

**Comment by danez** [comment]
We should keep an eye on performance with changes like this. I created a performance test-suite a while ago, so we should try this maybe. https://github.com/danez/babylon_performance

**Comment by xtuc** [comment]
Yes absolutely.

Branch `refactor-tokenizer-types`:

| name                            | run    |
|-|-|
| ./fixtures/angular.js           | 494ms  |
| ./fixtures/backbone.js          | 46ms   |
| ./fixtures/ember.debug.js       | 1223ms |
| ./fixtures/jquery.js            | 243ms  |
| ./fixtures/react-with-addons.js | 525ms  |

Branch `master`:

| name                            | run    |
|-|-|
| ./fixtures/angular.js           | 514ms  |
| ./fixtures/backbone.js          | 55ms   |
| ./fixtures/ember.debug.js       | 1262ms |
| ./fixtures/jquery.js            | 242ms  |
| ./fixtures/react-with-addons.js | 510ms  |


Diff:

| name                            | run    |
|-|-|
| ./fixtures/angular.js           | -18ms |
| ./fixtures/backbone.js          | -9ms   |
| ./fixtures/ember.debug.js       | -39ms |
| ./fixtures/jquery.js            | +1ms  |
| ./fixtures/react-with-addons.js | -15ms  |

Would be great to include your performance tests in Babylon tests (only on CI). What do you think about it?

**Comment by danez** [comment]
I thought about it, but wasn't sure how "stable" the performance on different boxes on travis is. But now that I think of it maybe if we would run the performance tests twice on travis in the same machine (for master and current commit/branch), the results could be useable.

File: src/tokenizer/types.js [hunk 1]
```diff
@@ -45,13 +58,13 @@ export const types = {
   eof: new TokenType("eof"),
 
   // Punctuation token types.
-  bracketL: new TokenType("[", {beforeExpr: true, startsExpr: true}),
+  bracketL: new TokenType("[", { ...beforeExpr, ...startsExpr }),
```
Comment (danez) [inline src/tokenizer/types.js]: Seems there are a lot of cases where we have `before` and `startsExpr` maybe we can do the spread in this case only once?

```js
const beforeAndStartExpr = { ...beforeExpr, ...startsExpr };
```

Comment (xtuc) [inline src/tokenizer/types.js]: Sure, of course. I will update it.

File: src/tokenizer/types.js [hunk 2]
```diff
@@ -32,10 +37,19 @@ export class TokenType {
   }
 }
 
-function binop(name, prec) {
-  return new TokenType(name, {beforeExpr: true, binop: prec});
+class KeyworkTokenType extends TokenType {
```
Comment (danharper) [inline src/tokenizer/types.js]: typo: `KeywordTokenType`

**Comment by xtuc** [comment]
@danharper I made some changes to improve the readability. I replaced the spread operators with property shorthands.

File: src/tokenizer/types.js [hunk 3]
```diff
@@ -32,38 +39,47 @@ export class TokenType {
...
-  colon: new TokenType(":", beforeExpr),
-  doubleColon: new TokenType("::", beforeExpr),
+  comma: new TokenType(",", { beforeExpr }),
+  semi: new TokenType(";", { beforeExpr }),
+  colon: new TokenType(":", { beforeExpr }),
+  doubleColon: new TokenType("::",{ beforeExpr }),
   dot: new TokenType("."),
-  question: new TokenType("?", beforeExpr),
-  arrow: new TokenType("=>", beforeExpr),
+  question: new TokenType("?",{ beforeExpr }),
```
Comment (danez) [inline src/tokenizer/types.js]: missing space after comma, seems we have no eslint rule active for that

File: src/tokenizer/types.js [hunk 4]
```diff
@@ -32,38 +39,47 @@ export class TokenType {
...
+  parenL: new TokenType("(", { beforeExpr, startsExpr }),
   parenR: new TokenType(")"),
-  comma: new TokenType(",", beforeExpr),
-  semi: new TokenType(";", beforeExpr),
-  colon: new TokenType(":", beforeExpr),
-  doubleColon: new TokenType("::", beforeExpr),
+  comma: new TokenType(",", { beforeExpr }),
+  semi: new TokenType(";", { beforeExpr }),
+  colon: new TokenType(":", { beforeExpr }),
+  doubleColon: new TokenType("::",{ beforeExpr }),
```
Comment (danez) [inline src/tokenizer/types.js]: same here

**Pull request merged**